Wrapper for GMM

kmetra1910 opened the pull request:
Drop-in replacement for GMM from gmr

Implements:

.from_samples(...)

.sample(...)

.predict(...)

.to_responsibilities(...)

.condition(...)

Carefully replicates gmr behavior, including:

Manual initialization when n_samples < 2

Safe handling of degenerate weights (e.g., nan or negative)

Adds warnings instead of exceptions for better debug experience

![image](https://github.com/user-attachments/assets/1bd3eb15-56b9-45db-bcb2-addb81cddd79)




This pull request includes several changes across multiple files to improve code readability and consistency. The most important changes involve adding missing imports, reformatting code for better readability, and updating string formatting.

### Code Readability Improvements:
* [`bamt/networks/base.py`](diffhunk://#diff-61db96ae3dc3f0a00450ce48cc1225a04aa7dbc0369dbfe160cd9c2d3e2257daL6-R17): Reformatted import statements and long lines to improve readability. [[1]](diffhunk://#diff-61db96ae3dc3f0a00450ce48cc1225a04aa7dbc0369dbfe160cd9c2d3e2257daL6-R17) [[2]](diffhunk://#diff-61db96ae3dc3f0a00450ce48cc1225a04aa7dbc0369dbfe160cd9c2d3e2257daL66-R79) [[3]](diffhunk://#diff-61db96ae3dc3f0a00450ce48cc1225a04aa7dbc0369dbfe160cd9c2d3e2257daL506-R519) [[4]](diffhunk://#diff-61db96ae3dc3f0a00450ce48cc1225a04aa7dbc0369dbfe160cd9c2d3e2257daL866-R879) [[5]](diffhunk://#diff-61db96ae3dc3f0a00450ce48cc1225a04aa7dbc0369dbfe160cd9c2d3e2257daL1018-R1043)
* [`bamt/nodes/conditional_mixture_gaussian_node.py`](diffhunk://#diff-21d883f24c00166c4434adb50bf076f144457b679cf3ffa3cd365ff3e66f5016L143-R147): Reformatted long lines and updated the `choose` and `predict` methods for better readability. [[1]](diffhunk://#diff-21d883f24c00166c4434adb50bf076f144457b679cf3ffa3cd365ff3e66f5016L143-R147) [[2]](diffhunk://#diff-21d883f24c00166c4434adb50bf076f144457b679cf3ffa3cd365ff3e66f5016L155-R161) [[3]](diffhunk://#diff-21d883f24c00166c4434adb50bf076f144457b679cf3ffa3cd365ff3e66f5016L183-R189) [[4]](diffhunk://#diff-21d883f24c00166c4434adb50bf076f144457b679cf3ffa3cd365ff3e66f5016L220-R232)

### Consistency Updates:
* [`bamt/external/pyitlib/DiscreteRandomVariableUtils.py`](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L633-R662): Updated string formatting to use double quotes and reformatted long lines for consistency. [[1]](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L633-R662) [[2]](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L682-R712) [[3]](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L725-R733) [[4]](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L738-R772)
* [`bamt/nodes/conditional_logit_node.py`](diffhunk://#diff-74d92151921e12bb06952832057a1010c57c56f24bd4e47b919c4ec352ae2895L95-R115): Reformatted long lines and updated string formatting for consistency.

### Minor Additions:
* Added missing author information in test files. [[1]](diffhunk://#diff-ef17a2ff7525efb817be4917399038226413c607f76585398b902de3458d5c43R34) [[2]](diffhunk://#diff-8654b2c03a0f2cb5824e8a5a97c570939385f356186aff73fb0014908509698fR8) [[3]](diffhunk://#diff-a670e8ff77a6b9ef527e1dabd7bf2243767817aa998d80b2ce0e682ef1ec4271R8) [[4]](diffhunk://#diff-66315e43bf2c9af80365ca8decf85f25daf74c4083f03ef3d205334afe94c526R8)

### Bug Fixes:
* [`bamt/external/pyitlib/DiscreteRandomVariableUtils.py`](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L633-R662): Fixed assertion statements and improved error handling in `_estimate_probabilities` function. [[1]](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L633-R662) [[2]](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L725-R733) [[3]](diffhunk://#diff-86cb24636dac6d6ca6cb4161b416a932aeee551cd7c43ae3fa6cec1194310ca2L738-R772)

### Import Updates:
* [`bamt/nodes/conditional_mixture_gaussian_node.py`](diffhunk://#diff-21d883f24c00166c4434adb50bf076f144457b679cf3ffa3cd365ff3e66f5016L5-R5): Replaced import from `gmr` to `bamt.utils.gmm_wrapper` for `GMM`.

kmetra1910 committed:
Create gmm_wrapper.py

kmetra1910 committed:
Твоё сообщение к коммиту

kmetra1910 committed:
initialisation to be as it is in gmr but using only sklearn mixture package

kmetra1910 committed:
Init as in gmr

kmetra1910 committed:
Init as in gmr

kmetra1910 committed:
Init as in gmr

kmetra1910 committed:
wouldnt have been complient with sklearn so changed from attributes to methods

kmetra1910 committed:
Added .from_samples() method. Unfortunately it yet not supports weights because of library restrictions.

kmetra1910 committed:
Added .sample with RE in case not initialised self._gmm

kmetra1910 committed:
Added .sample with RE in case not initialised self._gmm

kmetra1910 committed:
First testcasing for initial wrapper. Not consistent with order of added methods, but i checked method sample.

kmetra1910 committed:
added checks for to_responsibilities method

kmetra1910 committed:
added to_responsibilities

kmetra1910 committed:
Finnaly finished god forsaken condition

kmetra1910 committed:
Did unit-tests to see if gmr's GMM's condition works as our GMM's condition. It requires manual entry of params because we cant garantee that sklearn's from_samples and gmr's going to be the same

kmetra1910 committed:
updated to get parametrized n_components and n_features

kmetra1910 committed:
relevant parameters names for sklearn from samples

kmetra1910 committed:
method renamed in wrapper

kmetra1910 committed:
method renamed in wrapper

kmetra1910 committed:
edging

kmetra1910 committed:
edging

kmetra1910 committed:
warning and nans added in sampling

kmetra1910 committed:
final update

kmetra1910 committed:
PEP8 with black

jrzkaminski commented on bamt/utils/gmm_wrapper.py:
```diff
@@ -0,0 +1,209 @@
+            self._manual_init(means, covariances, priors)
+
+    def _manual_init(self, means, covariances, priors):
+        # Преобразуем всё в numpy-массивы
```
т.к. либа все-таки рассчитана на широкую аудиторию, комменты должны быть на английском

jrzkaminski commented on bamt/utils/gmm_wrapper.py:
```diff
@@ -0,0 +1,209 @@
+from scipy.stats import multivariate_normal
+
+
+class GMM:
```
В целом по коду не хватает очень docstrings и типизации

jrzkaminski commented on bamt/utils/gmm_wrapper.py:
```diff
@@ -0,0 +1,209 @@
+        priors=None,
+        means=None,
+        covariances=None,
+        random_state=None,
```
Пример типизации по новому стандарту питона (с версии 3.10)
```suggestion
        n_components: int = 1,
        priors: np.ndarray | None = None,
        means: np.ndarray | None = None,
        covariances: np.ndarray | None = None,
        random_state: int | None = None,
```

jrzkaminski commented on bamt/utils/gmm_wrapper.py:
```diff
@@ -0,0 +1,209 @@
+
+    @property
+    def means(self):
+        return self._gmm.means_
```
В проперти можно добавить какие-то такие error handling, @Roman223 вроде делал какие-то другие шаблоны Exception (необязательно использовать RuntimeError), это можно глянуть в одном из модулей

```suggestion
    @property
    def means(self) -> np.ndarray:
        """Return the component means."""
        if self._gmm is None:
            raise RuntimeError("GMM is not initialized.")
        return self._gmm.means_
```

kmetra1910 committed:
PR review updated

kmetra1910 commented:
Обновил исходя из PR review, добавил комменты, докстринги и остальное

jrzkaminski commented:
Разве что надо убрать gmr теперь из requirements и pyproject.toml

Roman223 reviewed (commented):
**NB**: I didn't run these changes on my own!

## Low-Level Observations

- The low-level implementation looks solid overall.
- A few commented-out lines should be removed to clean things up.
- Exception handling is clear and well-structured.
- Computation is efficient and numerically stable — e.g., good use of `np.linalg.solve` instead of inverse matrix operations.

## High-Level Issues

### Redundant Validations in `from_samples`

There's a manual check to ensure `X.shape[0] >= 2`, like:

```python
if X.shape[0] < 2:
    raise ValueError("Need at least 2 samples...")
```

But this check already exists within `sklearn.GaussianMixture` through:

Snippet from `fit_predict`
```python
X = validate_data(self, X, dtype=[np.float64, np.float32], ensure_min_samples=2)
```

Which internally uses:

```python
if ensure_min_samples > 0:
    n_samples = _num_samples(array)
    if n_samples < ensure_min_samples:
        raise ValueError(
            "Found array with %d sample(s) (shape=%s) while a"
            " minimum of %d is required%s."
            % (n_samples, array.shape, ensure_min_samples, context)
        )
```

**Recommendation**: These checks are redundant and can be removed or delegated to `sklearn`.

---

### `manual_init` Duplication

Instead of rewriting the initialization logic, it's more maintainable to reuse `sklearn`'s internal `_initialize` method. For reference, here's the original:

```python
def _initialize(self, X, resp):
    """Initialization of the Gaussian mixture parameters."""
    n_samples, _ = X.shape
    weights, means, covariances = None, None, None
    if resp is not None:
        weights, means, covariances = _estimate_gaussian_parameters(
            X, resp, self.reg_covar, self.covariance_type
        )
        if self.weights_init is None:
            weights /= n_samples

    self.weights_ = weights if self.weights_init is None else self.weights_init
    self.means_ = means if self.means_init is None else self.means_init

    if self.precisions_init is None:
        self.covariances_ = covariances
        self.precisions_cholesky_ = _compute_precision_cholesky(
            covariances, self.covariance_type
        )
    else:
        self.precisions_cholesky_ = _compute_precision_cholesky_from_precisions(
            self.precisions_init, self.covariance_type
        )
```

**Recommendation**: Subclass `GaussianMixture` and override only what you need, e.g.:

```python
from sklearn.mixture import GaussianMixture

class CustomGMM(GaussianMixture):
    def _initialize(self, X, resp):
        super()._initialize(X, resp)
        # Add any custom behavior here
```

This avoids code duplication and benefits from future improvements in scikit-learn.

##  Testing Strategy

- Using `gmr` as a reference is acceptable for transitional verification.
- However, the long-term goal is to **remove `gmr`** from `bamt`, which means the testing strategy should evolve.

**Suggestions:**

- Shift from comparison-based testing to **behavioral testing**:
  - Check that likelihood improves after training.
  - Verify cluster consistency on toy datasets.
- Add **integration tests** on synthetic datasets with known structure.

## Action Items

1. Remove commented-out lines and any redundant stability checks.
2. Refactor by subclassing `sklearn.GaussianMixture` and override only required methods.
3. Redesign tests to focus on correctness and invariants instead of output comparison with `gmr`.

Roman223 commented:
@jrzkaminski убирать тогда надо аккуратно -- перенести например в группу тестирования

kmetra1910 committed:
PR review 2 updated, delegatesample size check to sklearn

kmetra1910 commented:
Removed redundant validation from from_samples() method and cleaned up outdated comments.
Confirmed that sklearn.GaussianMixture already handles insufficient data with its internal validation logic (as @Roman223 suggested).

 All tests pass
Suggest moving gmr removal and behavioral test redesign into the next PR, along with a cleaner initialization approach (as suggested by Roman) using sklearn subclassing or controlled parameter setup.

This will allow us to fully drop dependency on gmr and transition to robust behavioral testing (log-likelihood checks, cluster validation).
